src/Microsoft.Health.Dicom.Core/Features/Retrieve/AcceptHeaderHandler.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using EnsureThat;
+using FellowOakDicom;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Core.Exceptions;
+using Microsoft.Health.Dicom.Core.Messages;
+using Microsoft.Health.Dicom.Core.Messages.Retrieve;
+using Microsoft.Health.Dicom.Core.Web;
+
+namespace Microsoft.Health.Dicom.Core.Features.Retrieve;
+
+public class AcceptHeaderHandler : IAcceptHeaderHandler
+{
+    protected internal static readonly IReadOnlyDictionary<ResourceType, List<AcceptHeaderDescriptor>>
+        AcceptableDescriptors =
+            new Dictionary<ResourceType, List<AcceptHeaderDescriptor>>()
+            {
+                { ResourceType.Study, DescriptorsForGetNonFrameResource(PayloadTypes.MultipartRelated) },
+                { ResourceType.Series, DescriptorsForGetNonFrameResource(PayloadTypes.MultipartRelated) },
+                { ResourceType.Instance, DescriptorsForGetNonFrameResource(PayloadTypes.SinglePartOrMultipartRelated) },
+                { ResourceType.Frames, DescriptorsForGetFrame() },
+            };
+
+    private readonly IReadOnlyDictionary<ResourceType, List<AcceptHeaderDescriptor>> _acceptableDescriptors;
+
+    private readonly ILogger<AcceptHeaderHandler> _logger;
+
+    public AcceptHeaderHandler(ILogger<AcceptHeaderHandler> logger)
+        : this(AcceptableDescriptors, logger)
+    {
+    }
+
+    public AcceptHeaderHandler(
+        IReadOnlyDictionary<ResourceType, List<AcceptHeaderDescriptor>> acceptableDescriptors,
+        ILogger<AcceptHeaderHandler> logger)
+    {
+        EnsureArg.IsNotNull(logger, nameof(logger));
+        EnsureArg.IsNotNull(acceptableDescriptors, nameof(acceptableDescriptors));
+
+        _acceptableDescriptors = acceptableDescriptors;
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// Based on requested AcceptHeaders from users ordered by priority, create new AcceptHeader with valid
+    /// TransferSyntax, leaving user input unmodified.
+    /// </summary>
+    /// <param name="resourceType">Used to understand if header properties are valid.</param>
+    /// <param name="acceptHeaders">One or more headers as requested by user.</param>
+    /// <returns>New accept header based on highest priority valid header requested.</returns>
+    /// <exception cref="NotAcceptableException"></exception>
+    public AcceptHeader GetValidAcceptHeader(ResourceType resourceType, IEnumerable<AcceptHeader> acceptHeaders)
+    {
+        EnsureArg.IsNotNull(acceptHeaders, nameof(acceptHeaders));
+        List<AcceptHeader> orderedHeaders = acceptHeaders.OrderByDescending(x => x.Quality).ToList();
+
+        _logger.LogInformation(
```

Comment:
It might be useful to have this log statement first, so that invalid headers would be logged before throwing an exception.